### ISSUE TITLE
Quaternion library is inconsistent.

### ISSUE THREAD

**Issue by Viech** [Unvanquished/Unvanquished#945]
Vec3 a = {1.0f,2.0f,3.0f}, b;
    quat_t aq;
    AnglesToQuat(a.Data(), aq);
    QuatToAngles(aq, b.Data());
    turretLogger.Debug("%s %s", Utility::Print(a), Utility::Print(b));

→ `Debug: (1.00, 2.00, 3.00) (1.00, 178.00, 177.00)`

**Comment by Viech** [comment]
The commist supposedly fixing the issue aren't on master anymore. Refer to #946 for a squashed version.

**Comment by DolceTriade** [comment]
One nit, and otherwise I'm fine with this change. We can either wait for that feature (probably never) or revert that bit of the change.

**Comment by Viech** [comment]
I may have found another inconsistency. The headers for the quaternion multiplication functions explain how they can be applied for rotations (see `q_shared.h`):

```
// qa = rotate by qa, then qb
	void QuatMultiply0( quat_t qa, const quat_t qb );

// qc = rotate by qa, then qb
	void QuatMultiply1( const quat_t qa, const quat_t qb, quat_t qc );

// qc = rotate by qa, then by inverse of qb
	void QuatMultiply2( const quat_t qa, const quat_t qb, quat_t qc );

// qc = rotate by inverse of qa, then by qb
	void QuatMultiply3( const quat_t qa, const quat_t qb, quat_t qc );

// qc = rotate by inverse of qa, then by inverse of qb
	void QuatMultiply4( const quat_t qa, const quat_t qb, quat_t qc );
```

I assumed that rotating by `a`, then by `b` and lastly by the inverse of `b` would yield `a` as a result. Following the description of the multiplication functions above, this does not seem to be the case:

```
	quat_t a; Vec3 aAngles = { 5.0f,10.0f,15.0f}; AnglesToQuat(aAngles.Data(), a);
	quat_t b; Vec3 bAngles = {90.0f,90.0f,90.0f}; AnglesToQuat(bAngles.Data(), b);
	quat_t aThenB, aThenBThenBInverse;
	QuatMultiply1(a, b, aThenB);
	QuatMultiply2(aThenB, b, aThenBThenBInverse);
	Vec3 aThenBAngles; QuatToAngles(aThenB, aThenBAngles.Data());
	Vec3 aThenBThenBInverseAngles; QuatToAngles(aThenBThenBInverse, aThenBThenBInverseAngles.Data());
	turretLogger.Debug("a: %s. b: %s. a then b: %s. a then b then inverse of b: %s.",
		Utility::Print(aAngles), Utility::Print(bAngles), Utility::Print(aThenBAngles),
		Utility::Print(aThenBThenBInverseAngles)
	);
```

→ `Debug: a: (5.00, 10.00, 15.00). b: (90.00, 90.00, 90.00). a then b: (74.21, 118.02, 15.11). a then b then inverse of b: (2.16, 180.00, 180.00).`

It seems that either my assumption, the function description or the function implementation is wrong. Note that I did not test the other multiplications. It may be that all of them have related issues.

**If you want to propose a fix, please add the commit to #946 for review!**

**Comment by Viech** [comment]
On the current quatfix branch, that is on 857c00d, there still seems to be something wrong with quats and angles.

Here's a testbench that mimics my turret code (download it as a patch for 857c00d [here](http://dl.viech.name/quatfix.patch)). The turret has a _torso orientation_ (relative to the world) and a _head orientation_ (relative to the torso) which together make for an _absolute orientation_ (of the turret's head relative to the world).

```
static const Vec3 AbsoluteAngles(const Vec3& torsoAngles, const Vec3& headAngles) {
	quat_t torsoRotation, headRotation;
	AnglesToQuat(torsoAngles.Data(), torsoRotation);
	AnglesToQuat(headAngles.Data(),  headRotation);

	// To get the turret's absolute rotation (relative to the world), first rotate by the torso
	// rotation (relative to the world), then by the head rotation (relative to the torso).
	quat_t absoluteRotation;
	QuatMultiply(headRotation, torsoRotation, absoluteRotation);

	vec3_t absoluteAngles;
	QuatToAngles(absoluteRotation, absoluteAngles);
	return Vec3::Load(absoluteAngles);
}

static const Vec3 HeadAngles(const Vec3& torsoAngles, const Vec3& absoluteAngles) {
	quat_t torsoRotation, absoluteRotation;
	AnglesToQuat(torsoAngles.Data(),    torsoRotation);
	AnglesToQuat(absoluteAngles.Data(), absoluteRotation);

	quat_t inverseTorsoRotation;
	QuatCopy(torsoRotation, inverseTorsoRotation);
	QuatInverse(inverseTorsoRotation);

	// To get the turret's head rotation (relative to the torso), first rotate by the inverse torso
	// rotation (relative to the world), then by the absolute rotation (relative to the world).
	quat_t headRotation;
	QuatMultiply(absoluteRotation, inverseTorsoRotation, headRotation);

	vec3_t headAngles;
	QuatToAngles(headRotation, headAngles);
	return Vec3::Load(headAngles);
}

static std::string P(const Vec3& v) {
	return std::string(va("(%7.2f, %7.2f, %7.2f)", v[0], v[1], v[2]));
}

static void DoQuatTests() {
	Log::Notice("------- Quaternion Test -------");

	const static std::pair<Vec3, Vec3> vecs[] = {
		// Pitch tests.
		{{   0,   0,   0}, {  90,   0,  0}},
		{{  90,   0,   0}, {  90,   0,  0}},
		{{  90,   0,   0}, { -90,   0,  0}},
		// Combined tests.
		{{   0,   0,   0}, {   0,  45,  0}},
		{{  90,   0,   0}, {   0,  45,  0}},
		{{ -90,   0,   0}, {   0,  45,  0}},
		{{ 180,   0,   0}, {   0,  45,  0}},
	};

	Vec3 t, h, a, x;

	for (int i = 0; i < sizeof(vecs) / sizeof(std::pair<Vec3, Vec3>); i++) {
		t = vecs[i].first;
		h = vecs[i].second;

		Log::Notice("t = %s, h = %s → a = %s.", P(t), P(h), P(a = AbsoluteAngles(t, h)));
		Log::Notice("t = %s, a = %s → h = %s.", P(t), P(a), P(x = HeadAngles(t, a)));
		Log::Notice("\nOld h: %s.\nNew h: %s.\n\n", P(h), P(x));
	}
}
```

This is the output:

```
------- Quaternion Test -------
t = (   0.00,    0.00,    0.00), h = (  90.00,    0.00,    0.00) → a = (  89.98,    0.00,    0.00).
t = (   0.00,    0.00,    0.00), a = (  89.98,    0.00,    0.00) → h = (  89.98,    0.00,    0.00).

Old h: (  90.00,    0.00,    0.00).
New h: (  89.98,    0.00,    0.00).


t = (  90.00,    0.00,    0.00), h = (  90.00,    0.00,    0.00) → a = (  -0.00,  180.00,    0.00).
t = (  90.00,    0.00,    0.00), a = (  -0.00,  180.00,    0.00) → h = ( -89.98,    0.00,  180.00).

Old h: (  90.00,    0.00,    0.00).
New h: ( -89.98,    0.00,  180.00).


t = (  90.00,    0.00,    0.00), h = ( -90.00,    0.00,    0.00) → a = (  -0.00,    0.00,    0.00).
t = (  90.00,    0.00,    0.00), a = (  -0.00,    0.00,    0.00) → h = ( -89.98,    0.00,    0.00).

Old h: ( -90.00,    0.00,    0.00).
New h: ( -89.98,    0.00,    0.00).


t = (   0.00,    0.00,    0.00), h = (   0.00,   45.00,    0.00) → a = (  -0.00,   45.00,    0.00).
t = (   0.00,    0.00,    0.00), a = (  -0.00,   45.00,    0.00) → h = (  -0.00,   45.00,    0.00).

Old h: (   0.00,   45.00,    0.00).
New h: (  -0.00,   45.00,    0.00).


t = (  90.00,    0.00,    0.00), h = (   0.00,   45.00,    0.00) → a = (   -nan,    0.00,    0.00).
t = (  90.00,    0.00,    0.00), a = (   -nan,    0.00,    0.00) → h = (   -nan,    -nan,    -nan).

Old h: (   0.00,   45.00,    0.00).
New h: (   -nan,    -nan,    -nan).


t = ( -90.00,    0.00,    0.00), h = (   0.00,   45.00,    0.00) → a = (   -nan,    0.00,    0.00).
t = ( -90.00,    0.00,    0.00), a = (   -nan,    0.00,    0.00) → h = (   -nan,    -nan,    -nan).

Old h: (   0.00,   45.00,    0.00).
New h: (   -nan,    -nan,    -nan).


t = ( 180.00,    0.00,    0.00), h = (   0.00,   45.00,    0.00) → a = (   0.00, -135.00,    0.00).
t = ( 180.00,    0.00,    0.00), a = (   0.00, -135.00,    0.00) → h = (  -0.00,   45.00,  180.00).

Old h: (   0.00,   45.00,    0.00).
New h: (  -0.00,   45.00,  180.00).
```

As you can see, given a fixed torso orientation, the `AbsoluteAngles` and `HeadAngles` are not inverse functions.

**Comment by Viech** [comment]
Fixed by merge of #946.
